### PR TITLE
Back-off and limits strategy for order-book API

### PR DESCRIPTION
**Pull request by shoom3301**
Reference: https://github.com/cowprotocol/cowswap/pull/2069

Related to:
 - https://github.com/cowprotocol/cow-sdk/pull/121
 - https://github.com/cowprotocol/cow-sdk/pull/120
 
 # What changed
1. Removed usage and generation of `OrderBookClient` (see package.json changes) because it doesn't match with `backOff`. Ignore changes in `generated` directory
2. Copied `backOff` and `rateLimiter` logic from CowSwap
3. Renamed an argument of `orderBookApi.getTrades`: `orderId` -> `orderUid`
4. Added docs to `README` about new options

### CODE REVIEW

File: src/order-book/api.spec.ts [hunk 1]
```diff
@@ -19,7 +19,11 @@ import { BUY_ETH_ADDRESS } from './transformOrder'
 
 enableFetchMocks()
 
-const orderBookApi = new OrderBookApi({ chainId: SupportedChainId.GNOSIS_CHAIN })
+const orderBookApi = new OrderBookApi({ chainId: SupportedChainId.GNOSIS_CHAIN }, undefined, {
+  numOfAttempts: 1,
```
Comment (shoom3301) [inline src/order-book/api.spec.ts:23]: Disabled retries for regular API tests (we have `request.test.ts` for testing bacOff)

**Review by anxolin — COMMENTED** [review]
> Removed usage and generation of OrderBookClient (see package.json changes) because it doesn't match with backOff. Ignore changes in generated directory

I don't understand, what doesn't match? Is not a pitty to remove the automatic generation, that was a great addition.

File: src/order-book/request.ts [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+}
+
+export interface FetchParams {
+  path: string
+  method: 'GET' | 'POST' | 'DELETE' | 'PUT'
+  body?: unknown
+  query?: URLSearchParams
+}
+
+const getResponseBody = async (response: Response): Promise<any> => {
```
Comment (anxolin) [inline src/order-book/request.ts:55]: its complaining here for the any

File: src/order-book/request.ts [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  if (response.status !== 204) {
+    try {
+      const contentType = response.headers.get('Content-Type')
+      if (contentType) {
+        const isJSON = contentType.toLowerCase().startsWith('application/json')
```
Comment (anxolin) [inline src/order-book/request.ts:60]: maybe move to function `isJson`

File: src/order-book/request.ts [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+    const response = await fetch(url, init)
+    const responseBody = await getResponseBody(response)
+
+    // Successful response
+    if (response.status >= 200 && response.status < 300) {
+      return responseBody
+    }
+
+    return Promise.reject(new OrderBookApiError(response, responseBody))
```
Comment (anxolin) [inline src/order-book/request.ts:100]: Wouldn't this have the problem we observe in prod? (we re-attempt on errors that wouldn't benefit on re-attempting)

**Review by anxolin — COMMENTED** [review]
Soft approve. I mean, is not too bad to remove the use of the generated services since at least the types of the endpoints are generated, however, was not nice to have? 

I thought the implementation you did where the backoff thingy would receive an arbitrary callback. Wouldn't this be usable with the original code?

File: src/order-book/request.ts (hunk 4 above)
Comment (shoom3301) [reply src/order-book/request.ts:100]: Checked this errors, works well now

**Comment by shoom3301** [comment]
@anxolin it's a great question, I've tried to do so, but there are typings problems, you can see it here:
https://github.com/cowprotocol/cow-sdk/pull/123

File: src/order-book/request.ts (hunk 3 above)
Comment (shoom3301) [reply src/order-book/request.ts:60]: Since we use it only once I moved the expression just to `if()`

File: src/order-book/request.ts (hunk 2 above)
Comment (shoom3301) [reply src/order-book/request.ts:55]: Thanks! Fixed

File: README.md [hunk 5]
```diff
@@ -95,6 +95,64 @@ async function main() {
...
+#### API urls configuration
+```typescript
+import { OrderBookApi } from '@cowprotocol/cow-sdk'
+
+const orderBookApi = new OrderBookApi({
+  chainId: SupportedChainId.GOERLI,
+  baseUrls: { // <-----
+    [SupportedChainId.MAINNET]: 'https://YOUR_ENDPOINT/mainnet',
+    [SupportedChainId.GNOSIS_CHAIN]: 'https://YOUR_ENDPOINT/xdai',
+    [SupportedChainId.GOERLI]: 'https://YOUR_ENDPOINT/goerli',
```
Comment (alfetopito) [inline README.md:123]: Again, why would I want to change the url?

File: src/order-book/api.ts [hunk 6]
```diff
@@ -46,47 +39,31 @@ export const ORDER_BOOK_STAGING_CONFIG: ApiBaseUrls = {
...
-  private servicePerNetwork: Record<string, OrderBookClient | null> = {}
 
-  constructor(context: PartialApiContext = {}) {
+  private rateLimiter: RateLimiter
+
+  constructor(
+    context: PartialApiContext = {},
+    limiterOpts: RateLimiterOpts = DEFAULT_LIMITER_OPTIONS,
+    public readonly backoffOpts: BackoffOptions = DEFAULT_BACKOFF_OPTIONS
+  ) {
```
Comment (alfetopito) [inline src/order-book/api.ts:51]: Why are these options not part of the `context` option obj?

I personally prefer to have one single obj parameter, so if you want to omit some parameter and only provide the last, you don't need to pass `undefined` for them

File: README.md [hunk 7]
```diff
@@ -95,6 +95,64 @@ async function main() {
...
+`OrderBookApi` - is a main tool for working with [CoW Protocol API](https://api.cow.fi/docs/#/).
+Since the API supports different networks and environments, there are some options to configure it.
+
+#### Environment configuration
+```typescript
+import { OrderBookApi } from '@cowprotocol/cow-sdk'
+
+const orderBookApi = new OrderBookApi({
+  chainId: SupportedChainId.GOERLI,
+  env: 'staging' // <-----
```
[the remark this answers is not in the document]
Comment (shoom3301) [reply README.md:110]: Added more info

File: README.md (hunk 5 above)
Comment (shoom3301) [reply README.md:123]: Added more info

File: src/order-book/api.ts (hunk 6 above)
Comment (shoom3301) [reply src/order-book/api.ts]: I though about it :) Thanks! Fixed

File: README.md [hunk 8]
```diff
@@ -95,6 +95,74 @@ async function main() {
...
+const backOffOpts: BackoffOptions = {
+  numOfAttempts: 5,
+  maxDelay: Infinity,
+  jitter: 'none',
+}
+
+const orderBookApi = new OrderBookApi(
+  {chainId: SupportedChainId.GOERLI},
+  limiterOpts,
+  backOffOpts
```
[the remark this answers is not in the document]
Comment (shoom3301) [reply README.md]: Thanks! Fixed

**Review by anxolin — APPROVED** [review]

**Review by alfetopito — APPROVED** [review]

**Pull request merged**